### ISSUE TITLE
model.predict return None

### ISSUE THREAD

**Issue by fbenti** [ultralytics/ultralytics#3024]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm having issues with the `predict` and `track` functions since they always return None.

```
cv2_image = self.cv2_bridge.imgmsg_to_cv2(msg)
cv2_image = cv2.cvtColor(cv2_image, cv2.COLOR_BGR2RGB)
results = self.model.predict(cv2_image, conf=0.1, classes=[32], show=False, stream=False, tracker=self.tracker)
print(results)
```

The output is always None.
```
0: 480x640 (no detections), 72.9ms
Speed: 0.8ms preprocess, 72.9ms inference, 0.7ms postprocess per image at shape (1, 3, 640, 640)
None
```

What could be the cause? The model is custom trained. If I provide a video stream in mp4 format, it works fine.

**Comment by glenn-jocher** [comment]
Hello @fbenti, 

The `predict` function of YOLOv8 should always return a results object, which is a dictionary-like structure containing all detected objects and their attributes. 

However, in this case, it seems like the object detection model is not detecting any object in the input image provided. This is shown in the output: "0: 480x640 (no detections), 72.9ms" which means no objects were found in the image.

One possible reason could be the low confidence threshold provided (0.1). You could try increasing the confidence value as it could be discarding some objects due to a low threshold. 

Another factor could be the classes parameter. If the image provided doesn't contain any objects of class 32, there is a possibility that the model might not detect anything. You could try changing the classes parameter or setting it to None to detect all classes.

If you provide a video stream or a video as input to the function, objects might be detected more easily due to the continuity of the frames.

Let me know if this helps or if you have any other questions.

**Comment by Laughing-q** [comment]
@fbenti I'm not able to reproduce the issue, even there's no detections just like your case.
![pic-selected-230606-1011-55](https://github.com/ultralytics/ultralytics/assets/61612323/1fbc6a42-75dd-4c8a-961f-41b2226c2cc2)
BTW I don't think the `predict` accepts a `tracker` arg, have you modified the function?

**Comment by fbenti** [comment]
@Laughing-q Yeah sorry my bad, it's a typo since I was trying using the `track` function before. Anyway even without it, it still returns `None`.

> Hello @fbenti,
> 
> The `predict` function of YOLOv8 should always return a results object, which is a dictionary-like structure containing all detected objects and their attributes.
> 
> However, in this case, it seems like the object detection model is not detecting any object in the input image provided. This is shown in the output: "0: 480x640 (no detections), 72.9ms" which means no objects were found in the image.
> 
> One possible reason could be the low confidence threshold provided (0.1). You could try increasing the confidence value as it could be discarding some objects due to a low threshold.
> 
> Another factor could be the classes parameter. If the image provided doesn't contain any objects of class 32, there is a possibility that the model might not detect anything. You could try changing the classes parameter or setting it to None to detect all classes.
> 
> If you provide a video stream or a video as input to the function, objects might be detected more easily due to the continuity of the frames.
> 
> Let me know if this helps or if you have any other questions.

@glenn-jocher Isn't the low confidence threshold simply supposed to return more detections? Even when the confidence is low, meaning that there will be lot of misclassification?

Anyway, I try to change these parameters, but in any case, the object returned is always `None`, even when there is object recognition, like below:

```
0: 480x640 2 ships, 91.9ms
Speed: 1.0ms preprocess, 91.9ms inference, 0.8ms postprocess per image at shape (1, 3, 640, 640)
None
```

Could the problem be related to the fact that the predict function is inside a loop? But I dont't see the reason for that. Basically, the function is inside a callback, so everytime I received a certain msg (which contain the image) I run the `predict` function.

**Comment by Laughing-q** [comment]
@fbenti From my above screenshot, I'm not able to reproduce the issue, can you tell me your package version? and it'll also be very helpful if you can provide more code block.

**Comment by glenn-jocher** [comment]
@fbenti, glad to hear that updating the package version has solved the previous issue. Regarding the `masks` attribute being `None`, it could be related to the default setting of the configuration file. By default, the `masks` attribute is not generated during inference. You might want to check whether the `yolo8_csp.yaml` configuration file specifies the `save_mids` parameter as true in both train and val sections. 

If it is set to true, it would generate both `masks` and `coords` attributes during inference. If you need to generate `masks` attribute during inference, you could set `save_mids: true` in the configuration file before training the model. 

Another possible cause for `masks` being `None` could be when the model has not been trained for instance segmentation and as such, does not contain the attributes `masks` and `coords`. Check if you have trained the model on instance segmentation.

Please let us know if the issue persists or if you have any further questions.

**Comment by Laughing-q** [comment]
@fbenti you need to use a instance segmentation model if you want to predict masks, i.e yolov8n-seg.pt.

**Comment by glenn-jocher** [comment]
You're welcome, @fbenti! Let us know if you have further questions or concerns.

Regarding the issue of `masks` attribute being `None`, this is because YOLOv8 is an object detection model and not an instance segmentation model. Predicting masks requires an instance segmentation model like yolov8n-seg.pt.

If you have any further questions or need further assistance, please do not hesitate to ask.

**Comment by chokevin8** [comment]
Hi, I'm not sure if this is the right place to ask since the thread is closed, but I'm running into a same problem as well. I've tried to use the smallest segmentation model yolov8n-seg.pt all the way to the largest one yolov8x-seg.pt, and I'm still getting no object detected and the mask attribute is none. During training & validation, the results aren't great, but they aren't absolutely terrible (about 0.45 mAP50), and the image that I'm testing is not any different from training and validation. (Plus I clearly see segmentation predictions being made during validation in the example images automatically generated). Could this possibly be a bug for the segmentation models where masks aren't being generated?

**Comment by Laughing-q** [comment]
@chokevin8 hi, I think you're right! the masks is None when there's no objects detected by model. If you got no detections then the masks in your case is None.

**Comment by chokevin8** [comment]
@Laughing-q Oh, I did not mean to say that, that's my mistake. I have seven objects (blood_vessels) detected, but still no masks. Below is some of the code/outputs:

**Code:**
infer_dataloader = DataLoader(CFG.test_paths, batch_size = 64, shuffle = False, pin_memory=True)
my_model = YOLO(CFG.model_path)
for tests in tqdm(infer_dataloader):
    predictions = list(my_model.predict(tests, device = 0,iou=0.7,agnostic_nms=False, save_conf=True, augment = False))

**Output:**
0: 512x512 7 blood_vessels, 11.3ms
Speed: 5.0ms preprocess, 11.3ms inference, 21.0ms postprocess per image at shape (1, 3, 512, 512)
100%|██████████| 1/1 [00:19<00:00, 19.88s/it]

But when I look at predictions (there is actually only one test image available, so ignore the batch_size for now), it gives me this:
[ultralytics.yolo.engine.results.Results object with attributes:
 boxes: ultralytics.yolo.engine.results.Boxes object
 keypoints: None
 keys: ['boxes']
 masks: None
...

This inference was made using the yolov8m.pt object detection model, by the way. Even though you guys may not be able to replicate my issue, could you try reproducing this issue with the object detection model to see if masks are outputed when objects are detected? Thank you!

**Comment by chokevin8** [comment]
By the way, the version I'm currently using is ultralytics==8.0.116.

**Comment by Laughing-q** [comment]
@chokevin8 There are two cases I can think of when masks is None:
1. the model is a segmentation model but there's no objects detected.
2. the model is **not** a segmentation model, masks would always be None whether there are objects or not.

From your description, it seems you are using a detection model but expect a predicted masks? which is the second case so you will not getting masks.
You need a segmentation model if you want to predict masks.
